addressing/README.md

```diff
@@ -0,0 +1,337 @@
+defined protocols and the length of their address components.
+
+### Composing multiaddrs
+
+As shown above, protocol addresses can be composed within a multiaddr in a way
+that mirrors the composition of protocols within a networking stack.
+
+The terms generally used to describe composition of multiaddrs are
+"encapsulation" and "decapsulation", and they essentially refer to adding and
+removing protocol components from a multiaddr, respectively.
+
+#### Encapsulation
+
+A protocol is said to be "encapsulated within" another protocol when data from
+an "inner" protocol is wrapped by another "outer" protocol, often by re-framing
+the data from the inner protocol into the type of packets, frames or datagrams
+used by the outer protocol.
+
+Some examples of protocol encapsulation are HTTP requests encapsulated within
+TCP/IP streams, or TCP segments themselves encapsulated within IP datagrams.
+
+The multiaddr format was designed so that addresses encapsulate each other in
+the same manner as the protocols that they describe. The result is an address
+that begins with the "outermost" layer of the network stack and works
+progressively "inward". For example, in the address `/ip4/7.7.7.7/tcp/80/ws`,
+the outermost protocol is IPv4, which encapsulates TCP streams, which in turn
+encapsulate WebSockets.
+
+All multiaddr implementations provide an `Encapsulate` method, which combines
+two multiaddrs into a composite. For example, `/ip4/7.7.7.7` can encapsulate
+`/tcp/42` to become `/ip4/7.7.7.7/tcp/42`.
+
+Note that no "sanity checking" is performed when encapsulating multiaddrs, and
+it is possible to create valid but unsound/dysfunctional multiaddrs like `/tcp/42/udp/42`
+through encapsulation.
+
+#### Decapsulation
+
+Decapsulation takes a composite multiaddr and removes an "inner"
+multiaddr from it, returning the result.
+
+For example, if we start with `/ip4/7.7.7.7/tcp/1234/ws` and decapsulate `/ws`,
+the result is `/ip4/7.7.7.7/tcp/1234`. 
+
+It's important to note that decapsulation returns the original multiaddr up
+to the last occurrence of the decapsulated multiaddr. This may remove more
+than just the decapsulated component itself if there are more protocols
+encapsulated within it. Using our example above, decapsulating either
+`/tcp/1234/ws` _or_ `/tcp/1234` from `/ip4/7.7.7.7/tcp/ws` will result in
+`/ip4/7.7.7.7`. This is unsurprising if you consider the utility of the
+`/ip4/7.7.7.7/ws` address that would result from simply removing the `tcp`
+component.
+
+### The p2p multiaddr
+
+libp2p defines the `p2p` multiaddr protocol, whose address component is the
+[peer id][peer-id-spec] of a libp2p peer. The text representation of a `p2p`
+multiaddr looks like this:
+
+```
+/p2p/QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N
+```
+
+Where `QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N` is the base58-encoded
+multihash of a peer's public key, also known as their peer id.
```

Comment:
```suggestion
Where `QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N` is the string representation
of a peer's peer ID derived from its public key.
```

Peer IDs don't have to be base58 encoded multihashes of the peer's public key. I would keep it generic as suggested above, referring to the "string representation" only, leaving the concrete details to the peer ID specification.